### DISCUSSION TITLE
gcc Apple Silicon missing AddressSanitizer/UBSanitizer libraries?

### DISCUSSION THREAD

**Discussion by dgazzoni** [Homebrew/discussions#3384] [Everyday usage]
I have a problem trying to use AddressSanitizer and/or UBSanitizer with Homebrew's gcc on Apple Silicon. It does not happen on Intel.

1. Install gcc 11 or 12: `brew install gcc` or `brew install gcc@12`
2. Create the following simple C program as `test.c`:

```
int main() {
    return 0;
}
```

3. Compile with AddressSanitizer and/or UBSanitizer:

```
gcc -fsanitize=address -o test test.c # also fails with gcc-12, or -fsanitize=undefined
```

4. Rather than a successful compile, the following errors are output:

```
ld: library not found for -lasan # or -lubsan if using -fsanitize=undefined
collect2: error: ld returned 1 exit status
```

Indeed, looking at the files on `/opt/homebrew/Cellar/gcc/11.3.0/lib/gcc/11` on my Apple Silicon Mac, there's no `libasan.dylib` or `libubsan.dylib`. On the Intel Mac, they are present in `/usr/local/Cellar/gcc/11.3.0/lib/gcc/11`.

I tried to look at the formulas for gcc, to see whether there was some kind of special-casing for either Apple Silicon or Intel, but I was unable to find any.

Can anyone else reproduce? Should I file a bug?

**Comment by SMillerDev** [comment]
GCC on ARM is a pre-release version, it isn't officially supported yet.

But if you're lucky @fxcoudert might be able to help you with this.

**Comment by fxcoudert** [comment]
I think the sanitizer runtime libraries are simply not supported on aarch64-apple-darwin at this point.

**Comment by ceciver** [comment]
hello @dgazzoni I'm running into the same problem, did you find a way how to use AddressSanitizer with gcc m1?

**Reply by fxcoudert** [reply]
> the sanitizer runtime libraries are simply not supported on aarch64-apple-darwin at this point

**Comment by vedanta28** [comment]
@fxcoudert I am able to use `clang++ -std=c++17 -o x x.cpp -fsanitize=address -g ` But, using g++ fails. Is this an unsolvable issue? Like will the support for this ever come in future?

**Reply by fxcoudert** [reply]
“Impossible to see, the future is.”

**Reply by jilljenn** [reply]
@vedanta28 Not sure the compilation option does anything at all actually

**Comment by samuelpmish** [comment]
I'm affected by this as well, using `GCC 13.1.0 aarch64-apple-darwin22`.

@fxcoudert do you have any information on how I might follow the development progress on GCC's support for aarch64-apple-darwin?

**Reply by carlocab** [reply]
See https://github.com/iains/gcc-13-branch and related repos from the same user.

**Comment by vedanta28** [comment]
@fxcoudert Actually, there is another issue too, which came to me after installing gcc 13.1.0. Some code which earlier used to compile is not compiling anymore. I am not sure if I should downgrade it or remain for the bugs to be fixed. More info here: https://codeforces.com/blog/entry/116459, https://stackoverflow.com/questions/76287256/declaring-array-of-large-size-of-vectorint-using-gcc-causes-error-changing-to

**Reply by fxcoudert** [reply]
I have verified the issue and filed it here: https://github.com/iains/gcc-darwin-arm64/issues/107

**Comment by mbUSC** [comment]
I'm also affected by this problem

**Comment by dpo** [comment]
I have the same issues with both the gcc-12 and gcc-13 toolchains, for both arm64 and intel architectures. I haven't tried older toolchains.

**Reply by Bo98** [reply]
The platforms where libsanitizer is supported is limited: https://github.com/iains/gcc-darwin-arm64/blob/master-wip-apple-si/libsanitizer/configure.tgt

It doesn't support arm64 at all on macOS, and now in addition does not support macOS Ventura or later on x86_64 (see https://github.com/iains/gcc-darwin-arm64/commit/e722a1f42b28092c9f709a3f758fc4fe57db32b0).
